### PR TITLE
fix(api): thermocycler hold_time bug

### LINKED ISSUE #6392: bug: Thermocycler: Two consecutive set_temps with same temps but different hold times result in error
# Steps to reproduce

Run a thermocycler protocol with these two consecutive steps:

```
thermocycler.set_block_temperature(95)
thermocycler.set_block_temperature(95, hold_time_seconds=20)
```
When protocol run reaches step 2, it results in the following error:
`Thermocycler driver set the block temp to T=95 & H=20 but status reads T=95.0 & H=15.0`

# Current behavior

When the protocol calls a `set_temperature()` on thermocycler, the thermocycler driver sends the set_temp command to the module and waits for 5seconds to verify that it reads the new set_temp values from the module. This is necessary to avoid getting into synchronization issues like the one in #6143. 
But it adds an additional issue that if the thermocycler was already at the target temp, then a consecutive set_temp with the same target as before will start counting the timer down immediately after set_temp. By the time the driver asks for verification from the module about its new target temperature & hold_time, the hold_time counter has already decreased and no longer the same as what it was set to. So in effect, the driver never reads the value that the module was originally set to and after waiting for 5 seconds, raises the error.

# Expected behavior

The driver should be able to decide smartly whether hold time has been updated. 
It might actually not be necessary to check for hold_time in `set_temperature`. (Will need testing)

### PR DESCRIPTION
**Pull request by sanni-t**
# Overview

Closes #6392 

# Changelog

- Thermocycler driver: modified the rule that checks if `hold_time` has been updated. It now checks if `hold_time` was *probably* updated, based on the closeness of `hold_time` value read to `hold_time` value set.
- Thermocycler HW control: updated `wait_for_hold` to account for the uncertainty introduced by above logic.

# Review requests

- Are there any edge cases you can think of that could break this logic?
- Is there a better way to make this fix? (without changing the firmware preferably)
- The driver tests that raise error are time-consuming since the error is raised after waiting for 5 seconds. Is there a way to minimize this delay? Or is it okay to let it take all that time?

### Another approach: 
Since a set_temperature function exits only when hold_time has turned 0, we can count on that behavior and only check if hold_time is non-zero for a new set_temperature. This will make things simpler but will cause a problem in the case where cached hold_time was somehow non-zero; but that should never happen if the api is used as intended, should it?

# Risk assessment

High for thermocycler since it touches a core function of the thermocycler

### CODE REVIEW

**Review by b-cooper — COMMENTED** [review]
This seems like an acceptable fix to the bug. 

There is a new side-effect where setting a temperature with a hold time between 0 and 5 has the potential to be off by 4 seconds in the worst case. 

For instance, setting temp to 35°C with a 4 second hold time will result in the api waiting for 9 seconds in the worst case.  This seems unlikely to occur as the TC would probably respond quicker than 5 seconds, and even if it does that is a very small margin of error. It might be worth noting this in documentation somewhere incase those few seconds are worth being aware of when writing a protocol.

**Comment by sanni-t** [comment]
> This seems like an acceptable fix to the bug.
> 
> There is a new side-effect where setting a temperature with a hold time between 0 and 5 has the potential to be off by 4 seconds in the worst case.
> 
> For instance, setting temp to 35°C with a 4 second hold time will result in the api waiting for 9 seconds in the worst case. This seems unlikely to occur as the TC would probably respond quicker than 5 seconds, and even if it does that is a very small margin of error. It might be worth noting this in documentation somewhere incase those few seconds are worth being aware of when writing a protocol.

Actually, setting 35C with 4 sec hold_time will have the api wait only 4 seconds. The driver will exit set_temperature immediately since 4 seconds lies in the fuzzy time gap. Then the hardware control will implement a 4 second delay.
(Or am I not understanding your case correctly?)

File: api/src/opentrons/hardware_control/modules/thermocycler.py [hunk 1]
```diff
@@ -158,7 +161,7 @@ async def _execute_cycles(self,
             self._current_cycle_index = rep + 1  # science starts at 1
             for step_idx, step in enumerate(steps):
                 await self._execute_cycle_step(step, step_idx, volume)
-                await self.wait_for_hold()
+                await self.wait_for_hold()  # What does this wait do?
```
Comment (sanni-t) [inline api/src/opentrons/hardware_control/modules/thermocycler.py:164]: @b-cooper Do you know why we require this hold?

File: api/src/opentrons/hardware_control/modules/thermocycler.py [hunk 2]
```diff
@@ -201,12 +204,20 @@ async def wait_for_temp(self):
...
         This method returns only when hold time has elapsed
         """
-        while self.hold_time != 0:
-            await asyncio.sleep(0.1)
+        # If hold time is within the HOLD_TIME_FUZZY_SECONDS time gap, then,
+        # because of the driver's status poller delays, it is impossible to
+        # know for certain if self.hold_time holds the most recent value.
+        # So instead of counting on the cached self.hold_time, it is better to
+        # just wait for hold_time time
+        if 0 < hold_time <= HOLD_TIME_FUZZY_SECONDS:
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/modules/thermocycler.py:216]: This logic looks a bit weird. If `hold_time` is precisely 0 it will fail the check on line 216 and go to the `else` where it will immediately end the `while` loop.

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Definitely a logic error in there.

It's also kind of gross to me; what this is doing is working around us not having all the events we require for robust operation. What we need is for the thermocycler firmware to tell us when it acknowledged and is executing a command.

File: api/src/opentrons/hardware_control/modules/thermocycler.py (hunk 2 above)
Comment (sanni-t) [reply api/src/opentrons/hardware_control/modules/thermocycler.py:216]: Not really an error since we don't want to wait when hold_time is 0. But I understand if you mean I should instead make it 
```
if 0 <= hold_time <= HOLD_TIME_FUZZY_SECONDS:
```

Comment (sfoster1) [reply api/src/opentrons/hardware_control/modules/thermocycler.py:216]: Yeah it would be good for it to be explicitly caught, or at least documented. As it is it looks like a bug

**Review by mcous — APPROVED** [review]
Reviewed in person, LGTM

**Review by amitlissack — APPROVED** [review]

**Review by sfoster1 — APPROVED** [review]
Looks good to me

**Review by b-cooper — APPROVED** [review]
📦
